File: license-backend/src/license/license-token-verifier.ts
import { base64UrlDecode } from "./base64url";
import { TOKEN_ALGORITHM, TOKEN_TYPE, TOKEN_VERSION, type LicenseTokenClaims } from "./license-token";
import { TrustedLicenseSigningKeys } from "./trusted-signing-keys";

const MAX_TOKEN_CHARS = 16_384;
const MAX_HEADER_CHARS = 2_048;
const MAX_PAYLOAD_CHARS = 8_192;
const MAX_SIGNATURE_CHARS = 4_096;

export type TokenVerificationFailure = "TOKEN_INVALID" | "TOKEN_EXPIRED" | "UNKNOWN_KEY";
export type TokenVerificationResult =
  | { readonly ok: true; readonly claims: LicenseTokenClaims }
  | { readonly ok: false; readonly code: TokenVerificationFailure };

export class LicenseTokenVerifier {
  private constructor(private readonly keys: TrustedLicenseSigningKeys) {}

  static async create(configuredKeys: ReadonlyMap<string, string>): Promise<LicenseTokenVerifier> {
    const entries = Array.from(configuredKeys, ([kid, pem]) => {
      const match = /^-----BEGIN PUBLIC KEY-----\s+([A-Za-z0-9+/=\s]+?)\s+-----END PUBLIC KEY-----\s*$/u.exec(pem.trim());
      if (match?.[1] === undefined) throw new Error("Verification key PEM is malformed.");
      return { kid, algorithm: "RS256", spkiBase64: match[1].replace(/\s/gu, "") };
    });
    return LicenseTokenVerifier.fromRegistry(await TrustedLicenseSigningKeys.fromJson(JSON.stringify(entries)));
  }

  static fromRegistry(keys: TrustedLicenseSigningKeys): LicenseTokenVerifier { return new LicenseTokenVerifier(keys); }

  async verify(encoded: string, now: number): Promise<TokenVerificationResult> {
    if (encoded.length < 1 || encoded.length > MAX_TOKEN_CHARS) return invalid("TOKEN_INVALID");
    const segments = encoded.split(".");
    if (segments.length !== 3 || segments.some((segment) => segment.length === 0 || segment.includes("="))) {
      return invalid("TOKEN_INVALID");
    }
    const [headerSegment, payloadSegment, signatureSegment] = segments as [string, string, string];
    if (headerSegment.length > MAX_HEADER_CHARS || payloadSegment.length > MAX_PAYLOAD_CHARS || signatureSegment.length > MAX_SIGNATURE_CHARS) {
      return invalid("TOKEN_INVALID");
    }
    let header: Record<string, unknown>;
    try { header = parseObject(headerSegment); } catch { return invalid("TOKEN_INVALID"); }
    if (header.alg !== TOKEN_ALGORITHM || header.typ !== TOKEN_TYPE || typeof header.kid !== "string") return invalid("TOKEN_INVALID");
    const key = this.keys.resolve(header.kid);
    if (key === undefined) return invalid("UNKNOWN_KEY");
    let signature: Uint8Array;
    try { signature = base64UrlDecode(signatureSegment, MAX_SIGNATURE_CHARS); } catch { return invalid("TOKEN_INVALID"); }
    const valid = await crypto.subtle.verify(
      "RSASSA-PKCS1-v1_5", key, signature.slice().buffer as ArrayBuffer,
      new TextEncoder().encode(`${headerSegment}.${payloadSegment}`),
    );
    if (!valid) return invalid("TOKEN_INVALID");
    let payload: Record<string, unknown>;
    try { payload = parseObject(payloadSegment); } catch { return invalid("TOKEN_INVALID"); }
    const claims = parseClaims(payload);
    if (claims === null) return invalid("TOKEN_INVALID");
    if (now >= claims.expiresAtEpochSeconds) return invalid("TOKEN_EXPIRED");
    if (claims.issuedAtEpochSeconds > now) return invalid("TOKEN_INVALID");
    return { ok: true, claims };
  }
}

function parseObject(segment: string): Record<string, unknown> {
  const value: unknown = JSON.parse(new TextDecoder().decode(base64UrlDecode(segment)));
  if (typeof value !== "object" || value === null || Array.isArray(value)) throw new Error("Expected object.");
  return value as Record<string, unknown>;
}

function parseClaims(value: Record<string, unknown>): LicenseTokenClaims | null {
  const strings = ["licenseId", "deviceId", "installationId", "deviceHash", "packageName"] as const;
  if (strings.some((name) => typeof value[name] !== "string" || (value[name] as string).length === 0)) return null;
  const numbers = ["issuedAtEpochSeconds", "expiresAtEpochSeconds", "offlineValidUntilEpochSeconds"] as const;
  if (numbers.some((name) => !Number.isSafeInteger(value[name]))) return null;
  if (value.tokenVersion !== TOKEN_VERSION) return null;
  return value as unknown as LicenseTokenClaims;
}

function invalid(code: TokenVerificationFailure): TokenVerificationResult { return { ok: false, code }; }

File: app/src/main/java/com/trancong/dexworkspacetouch/license/activation/LicenseTokenVerifier.kt
package com.trancong.dexworkspacetouch.license.activation

import com.trancong.dexworkspacetouch.license.domain.LicenseDeviceIdentity
import com.trancong.dexworkspacetouch.license.domain.LicensePolicy
import com.trancong.dexworkspacetouch.license.domain.LicenseTimeProvider
import com.trancong.dexworkspacetouch.license.domain.LicenseTokenClaims
import com.trancong.dexworkspacetouch.license.domain.LicensedApplicationIdentity
import org.json.JSONObject
import java.security.KeyFactory
import java.security.PublicKey
import java.security.Signature
import java.security.spec.X509EncodedKeySpec
import java.security.interfaces.RSAPublicKey
import java.util.Base64

class TrustedLicenseSigningKeys private constructor(private val keys: Map<String, PublicKey>) {
    operator fun get(kid: String): PublicKey? = keys[kid]
    val size: Int get() = keys.size

    companion object {
        private val KID_PATTERN = Regex("^[A-Za-z0-9][A-Za-z0-9._-]{0,63}$")

        fun emptyForDebug(): TrustedLicenseSigningKeys = TrustedLicenseSigningKeys(emptyMap())

        fun fromRegistryJson(json: String): TrustedLicenseSigningKeys {
            require(json.isNotBlank()) { "Trusted signing-key registry is empty." }
            val entries = org.json.JSONArray(json)
            require(entries.length() > 0) { "Trusted signing-key registry is empty." }
            val configured = LinkedHashMap<String, String>()
            repeat(entries.length()) { index ->
                val entry = entries.getJSONObject(index)
                require(entry.length() == 3 && entry.has("kid") && entry.has("algorithm") && entry.has("spkiBase64")) {
                    "Trusted signing-key entry has unsupported fields."
                }
                val kid = entry.getString("kid")
                require(KID_PATTERN.matches(kid)) { "Trusted signing-key ID is invalid." }
                require(entry.getString("algorithm") == "RS256") { "Trusted signing-key algorithm is unsupported." }
                require(configured.put(kid, entry.getString("spkiBase64")) == null) { "Duplicate trusted signing-key ID." }
            }
            return fromBase64Spki(configured)
        }

        fun fromBase64Spki(keys: Map<String, String>): TrustedLicenseSigningKeys {
            require(keys.isNotEmpty()) { "Trusted signing-key registry is empty." }
            val factory = KeyFactory.getInstance("RSA")
            val parsed = keys.mapValues { (kid, value) ->
                require(KID_PATTERN.matches(kid)) { "Trusted signing-key ID is invalid." }
                val decoded = Base64.getDecoder().decode(value)
                val key = factory.generatePublic(X509EncodedKeySpec(decoded)) as? RSAPublicKey
                    ?: error("Trusted signing key is not RSA.")
                require(key.modulus.bitLength() >= 2048) { "Trusted RSA signing key is smaller than 2048 bits." }
                key
            }
            return TrustedLicenseSigningKeys(parsed)
        }
    }
}

fun interface LicenseTokenValidation {
    fun verify(
        encoded: String,
        device: LicenseDeviceIdentity,
        application: LicensedApplicationIdentity,
    ): LicenseTokenVerificationResult
}

class LicenseTokenVerifier(
    private val trustedKeys: TrustedLicenseSigningKeys,
    private val timeProvider: LicenseTimeProvider,
) : LicenseTokenValidation {
    override fun verify(
        encoded: String,
        device: LicenseDeviceIdentity,
        application: LicensedApplicationIdentity,
    ): LicenseTokenVerificationResult {
        if (encoded.length !in 1..MAX_TOKEN_CHARS) return invalid(LicenseTokenInvalidReason.MALFORMED)
        val segments = encoded.split('.')
        if (segments.size != 3 || segments.any(String::isEmpty)) return invalid(LicenseTokenInvalidReason.MALFORMED)
        if (segments[0].length > MAX_HEADER_CHARS || segments[1].length > MAX_PAYLOAD_CHARS ||
            segments[2].length > MAX_SIGNATURE_CHARS || segments.any { '=' in it }) {
            return invalid(LicenseTokenInvalidReason.MALFORMED)
        }
        val decoder = Base64.getUrlDecoder()
        val header = try { JSONObject(String(decoder.decode(segments[0]), Charsets.UTF_8)) }
        catch (_: IllegalArgumentException) { return invalid(LicenseTokenInvalidReason.MALFORMED) }
        catch (_: org.json.JSONException) { return invalid(LicenseTokenInvalidReason.MALFORMED) }
        if (header.optString("alg") != "RS256" || header.optString("typ") != "DWT-LICENSE") {
            return invalid(LicenseTokenInvalidReason.UNSUPPORTED_HEADER)
        }
        val kid = header.optString("kid")
        val key = trustedKeys[kid] ?: return invalid(LicenseTokenInvalidReason.UNKNOWN_KEY)
        val signatureBytes = try { decoder.decode(segments[2]) }
        catch (_: IllegalArgumentException) { return invalid(LicenseTokenInvalidReason.MALFORMED) }
        val signatureValid = try {
            Signature.getInstance("SHA256withRSA").run {
                initVerify(key)
                update("${segments[0]}.${segments[1]}".toByteArray(Charsets.US_ASCII))
                verify(signatureBytes)
            }
        } catch (_: java.security.GeneralSecurityException) { false }
        if (!signatureValid) return invalid(LicenseTokenInvalidReason.INVALID_SIGNATURE)

        val payload = try { JSONObject(String(decoder.decode(segments[1]), Charsets.UTF_8)) }
        catch (_: IllegalArgumentException) { return invalid(LicenseTokenInvalidReason.MALFORMED) }
        catch (_: org.json.JSONException) { return invalid(LicenseTokenInvalidReason.MALFORMED) }
        val verified = try {
            val claims = LicenseTokenClaims(
                licenseId = payload.requireString("licenseId"),
                installationId = payload.requireString("installationId"),
                issuedAtEpochSeconds = payload.requireLong("issuedAtEpochSeconds"),
                expiresAtEpochSeconds = payload.requireLong("expiresAtEpochSeconds"),
                offlineValidUntilEpochSeconds = payload.requireLong("offlineValidUntilEpochSeconds"),
                packageName = payload.requireString("packageName"),
                tokenVersion = payload.requireInt("tokenVersion"),
            )
            VerifiedLicenseToken(encoded, claims, payload.requireString("deviceId"), payload.requireString("deviceHash"))
        } catch (_: Exception) { return invalid(LicenseTokenInvalidReason.INVALID_CLAIMS) }
        val claims = verified.claims
        if (claims.tokenVersion != LicensePolicy.TOKEN_VERSION || claims.issuedAtEpochSeconds > timeProvider.nowEpochMillis() / 1_000L) {
            return invalid(LicenseTokenInvalidReason.INVALID_CLAIMS)
        }
        if (timeProvider.nowEpochMillis() / 1_000L >= claims.expiresAtEpochSeconds) return invalid(LicenseTokenInvalidReason.TOKEN_EXPIRED)
        if (claims.installationId != device.installationId || verified.deviceHash != device.deviceHash) {
            return invalid(LicenseTokenInvalidReason.DEVICE_MISMATCH)
        }
        if (claims.packageName != application.packageName) return invalid(LicenseTokenInvalidReason.APPLICATION_MISMATCH)
        return LicenseTokenVerificationResult.Valid(verified)
    }

    private fun JSONObject.requireString(name: String): String = (get(name) as? String).orEmpty()
        .also { require(it.isNotBlank()) }
    private fun JSONObject.requireLong(name: String): Long {
        val value = get(name) as? Number ?: error("$name must be numeric")
        val result = value.toLong()
        require(value.toDouble() == result.toDouble()) { "$name must be an integer" }
        return result
    }
    private fun JSONObject.requireInt(name: String): Int {
        val value = requireLong(name)
        require(value in Int.MIN_VALUE..Int.MAX_VALUE) { "$name is out of range" }
        return value.toInt()
    }
    private fun invalid(reason: LicenseTokenInvalidReason) = LicenseTokenVerificationResult.Invalid(reason)

    private companion object {
        const val MAX_TOKEN_CHARS = 16_384
        const val MAX_HEADER_CHARS = 2_048
        const val MAX_PAYLOAD_CHARS = 8_192
        const val MAX_SIGNATURE_CHARS = 4_096
    }
}
